### PR TITLE
Mallocng

### PR DESCRIPTION
**Pull request by Maqrkk**
Mallocng should be compatible with cheerp now.

### CODE REVIEW

File: src/malloc/mallocng/glue.h [hunk 1]
```diff
@@ -39,55 +39,94 @@
...
 int __malloc_lock[1]; \
 void __malloc_atfork(int who) { malloc_atfork(who); }
 
+#ifdef __CHEERP__
+__attribute__((cheerp_asmjs))
+#endif
 static inline void rdlock()
 {
+#if 1
+#else
```
Comment (yuri91) [inline src/malloc/mallocng/glue.h:78]: what about these #if 1?

File: src/malloc/mallocng/meta.h [hunk 2]
```diff
@@ -6,22 +6,47 @@
...
-struct group {
-	struct meta *meta;
-	unsigned char active_idx:5;
+struct
+#ifdef __CHEERP__
+__attribute__((cheerp_asmjs))
+#endif
+group {
+	struct
+#ifdef __CHEERP__
```
Comment (yuri91) [inline src/malloc/mallocng/meta.h:36]: I *think* that the attribute in the inner struct is not required, since the enclosing context is the outer struct with the attribute. could you try if it's the case?

Comment (Maqrkk) [reply src/malloc/mallocng/meta.h:36]: I think because there's no forward declaration, it's seeing this as the declaration. It does not compile without the attribute here.

Comment (yuri91) [reply src/malloc/mallocng/meta.h:36]: ok

File: src/malloc/mallocng/glue.h (hunk 1 above)
Comment (Maqrkk) [reply src/malloc/mallocng/glue.h:78]: I noticed this construction in the oldmalloc. I thought since we're not using threading I could apply it here aswell. They did not compile at the time, due to MT (which I changed later). They do currently compile when I remove these if blocks.
Should I remove them?

Comment (yuri91) [reply src/malloc/mallocng/glue.h:78]: I would remove these if it compiles. They will still disappear in genericjs (since the MT=0). For wasm, it will make the code a bit worse, hopefully not much. You could keep them in a separate commit in a branch outside of the PR. Maybe Carlo could check if it makes a noticeable difference.
Eventually with threads we will need this code though....

Comment (Maqrkk) [reply src/malloc/mallocng/glue.h]: Removed the if blocks

File: src/malloc/mallocng/glue.h [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 #include <stdint.h>
 #include <sys/mman.h>
-#include <pthread.h>
+//#include <pthread.h>
```
Comment (yuri91) [inline src/malloc/mallocng/glue.h:6]: is this still needed?

File: src/malloc/mallocng/aligned_alloc.c [hunk 4]
```diff
@@ -1,7 +1,11 @@
+#define INTERNAL_MUSL
```
Comment (alexp-sssup) [inline src/malloc/mallocng/aligned_alloc.c:1]: Why are these defines needed?

File: src/malloc/mallocng/glue.h [hunk 5]
```diff
@@ -20,9 +20,9 @@
 #define is_allzero __malloc_allzerop
 #define dump_heap __dump_heap
 
-#define malloc __libc_malloc_impl
-#define realloc __libc_realloc
-#define free __libc_free
+//#define malloc __libc_malloc_impl
+//#define realloc __libc_realloc
+//#define free __libc_free
```
Comment (alexp-sssup) [inline src/malloc/mallocng/glue.h:25]: Is this required? If so just remove the lines, there is no reason to keep commented dead code around

File: src/malloc/mallocng/realloc.c [hunk 6]
```diff
@@ -31,7 +40,7 @@ void *realloc(void *p, size_t n)
 		size_t base = (unsigned char *)p-start;
 		size_t needed = (n + base + UNIT + IB + 4095) & -4096;
 		new = g->maplen*4096UL == needed ? g->mem :
-			mremap(g->mem, g->maplen*4096UL, needed, MREMAP_MAYMOVE);
+			mremap_wrapper(g->mem, g->maplen*4096UL, needed, MREMAP_MAYMOVE);
```
Comment (alexp-sssup) [inline src/malloc/mallocng/realloc.c:43]: Why is this required?

File: src/malloc/mallocng/glue.h [hunk 7]
```diff
@@ -51,7 +51,11 @@ static inline uint64_t get_random_secret()
 #define PAGESIZE PAGE_SIZE
 #endif
 
+#if defined(__CHEERP__) && !defined(__ASMJS__)
+#define MT (0)
```
Comment (alexp-sssup) [inline src/malloc/mallocng/glue.h:55]: Why only for genericjs?

File: src/malloc/mallocng/aligned_alloc.c (hunk 4 above)
Comment (Maqrkk) [reply src/malloc/mallocng/aligned_alloc.c:1]: In the /include/stdlib.h header file the malloc functions are given the cheerp_asmjs attribute only if INTERNAL_MUSL is defined. Otherwise they don't compile due to incompatible attributes.
This define is also present in oldmalloc.

Comment (alexp-sssup) [reply src/malloc/mallocng/aligned_alloc.c:1]: I am not convinced this thing make sense. Is there ever a case in which malloc and friend should _not_ be tagged? Please discuss with yuri

File: src/malloc/mallocng/glue.h (hunk 7 above)
Comment (yuri91) [reply src/malloc/mallocng/glue.h:55]: it could be for everything, but I thought that since we will add threads soon, it's better if the code compiles and works with this stuff

File: src/malloc/mallocng/glue.h (hunk 5 above)
Comment (yuri91) [reply src/malloc/mallocng/glue.h:25]: I left it like this throughout musl, because I felt that there should be a better way, and to avoid those defines coming back sneakily during a rebase. the better thing would actually be a #ifndef __CHEERP__ around them

File: src/malloc/mallocng/aligned_alloc.c (hunk 4 above)
Comment (yuri91) [reply src/malloc/mallocng/aligned_alloc.c:1]: it's to make it work when compiling with the genericjs target but mallocing asmjs structs

File: src/malloc/mallocng/realloc.c (hunk 6 above)
Comment (yuri91) [reply src/malloc/mallocng/realloc.c:43]: the function is variadic, and when compiling in genericjs the syscalls are genericjs, and it can't be used in malloc stuff because that is always asmjs. An alternative is to make it not variadic if the uses are all with the same arguments

Comment (alexp-sssup) [reply src/malloc/mallocng/realloc.c:43]: Let's go for not variadic.

File: src/malloc/mallocng/glue.h (hunk 3 above)
Comment (Maqrkk) [reply src/malloc/mallocng/glue.h:6]: It does not seem necessary anymore, will uncomment it

File: src/malloc/mallocng/realloc.c (hunk 6 above)
Comment (Maqrkk) [reply src/malloc/mallocng/realloc.c]: Pushed a new commit to make mremap not variadic.

File: src/malloc/mallocng/glue.h [hunk 8]
```diff
@@ -44,11 +44,7 @@ __attribute__((cheerp_asmjs))
 #endif
 static inline uint64_t get_random_secret()
 {
-	uint64_t secret = (uintptr_t)&secret * 1103515245;
-	for (size_t i=0; libc.auxv[i]; i+=2)
-		if (libc.auxv[i]==AT_RANDOM)
-			memcpy(&secret, (char *)libc.auxv[i+1]+8, sizeof secret);
-	return secret;
+	return 6148914691236517205; // represented in binary as alternating 0 and 1.
```
Comment (alexp-sssup) [inline src/malloc/mallocng/glue.h:47]: Add the `UL` postfix to the number, just to be on the safe side

Comment (Maqrkk) [reply src/malloc/mallocng/glue.h]: Done

File: src/malloc/mallocng/glue.h (hunk 5 above)
Comment (Maqrkk) [reply src/malloc/mallocng/glue.h]: I changed this to Yuri's suggestion.

File: src/malloc/mallocng/malloc.c [hunk 9]
```diff
@@ -292,6 +292,7 @@ static struct meta *alloc_group(int sc, size_t req)
 			return 0;
 		}
 		struct meta *g = ctx.active[j];
+		*(char*)(g->mem->storage + get_stride(g) * idx - 4) = 0;
```
Comment (yuri91) [inline src/malloc/mallocng/malloc.c:295]: Add a coment like:
```
// CHEERP: zero-initialize this location, since the code assumes that memory returned by mmap is zero-initialized, but in our case it's not
```

same in the other place below

Comment (Maqrkk) [reply src/malloc/mallocng/malloc.c:297]: Done.

**Review by yuri91 — APPROVED** [review]

**Review by yuri91 — APPROVED** [review]
